### ISSUE TITLE
Release Planning and pip/PyPI

### ISSUE THREAD

**Issue by lanctot** [deepmind/open_spiel#409]
Hi all, 

Just a quick heads-up that one of our big priorities is the planning another release, followed by finally getting OpenSpiel into PyPI so it can be installed quickly via pip (the team knows well by now that I love one-line installs!)

I am hoping the release will include binary distributions in shared object form (for MacOS and Linux x86_64), which will mean being able to use OpenSpiel as a library using only its headers via dynamic linking. This will save on compilation time and make it easier to use as a library rather than a framework.

Once this release is done, we'd like to get OpenSpiel into PyPI. This one is a bit trickier since the python API dynamically loads C++-compiled code as a Python extension. We will need to be sure we can support a wide enough set of platforms and python versions in order to make it widely accessible, and a good testing framework to ensure that it's stable. Still, I think it will be worth the effort. If anybody has experience doing this, please by all means let us know. 

Ideally the pair would arrive as a Christmas gift, but I think it is slightly optimistic to have both done by then. I think getting the release out by then is attainable... PyPI might take a bit longer, but we'll try for both.  Looking forward to closing this especially tough year with some good news, and will keep this thread for updates and discussion. Stay tuned!

**Comment by findmyway** [comment]
Hi @lanctot ,

Would you mind registering the Julia wrapper as a package under this repo directly? It was placed in https://github.com/JuliaReinforcementLearning/OpenSpiel.jl and most of the codes were just directly copied from the `julia` subfolder. But that's merely due to a restriction from the Julia packaging system. Several months ago, that restriction was addressed and now packages can be registered under the subfolder inside of a repo. The advantages are:

1. I don't need to maintain the same code in two different projects.
1. The julia package building processes are more transparent to users/developers.
1. The version number of the julia wrapper can be aligned with other wrappers. (It was wrongly tagged in the previous version.)

**Comment by lanctot** [comment]
> Would you mind registering the Julia wrapper as a package under this repo directly? 

Sounds good to me, but I don't know what is involved nor what is required of me.  Can you tell me more about what I have to do to enable this? And: can you ensure continued support it?

**Comment by findmyway** [comment]
> Can you tell me more about what I have to do to enable this?

Most things will be done automatically. The only thing needed is to update the version number here when you want to make a new release:

https://github.com/deepmind/open_spiel/blob/master/open_spiel/julia/Project.toml#L4

I'll make a PR to set up the pipeline and describe more details there.

> can you ensure continued support it?

Yes, I will.

**Comment by lanctot** [comment]
Ok, coming back to this now. 

Quick updates:

- I am unsure if I'll be able to release a binary distribution before the new year because of some interactions with abseil that I didn't anticipate (and having to test on many systems). I will still try for it, but might run out of time, and want to issue a release in December even if this does not make it.

- Despite that, I am looking into putting pyspiel into PyPI _as a source distribution, for now_ because the maintenance burden of a bdist wheel is too great right now. But, I'll still look into it over the year next year.

That said: the release that the pip package is based on would include some of the optional dependencies, basically all the ones that are license-compliant and easy to include just by re-distributing the source code, and so the global variables would be set to:

```
BUILD_WITH_PYTHON=ON
BUILD_WITH_HANABI=ON
BUILD_WITH_ACPC=ON
BUILD_WITH_EIGEN=ON
BUILD_WITH_PUBLIC_STATES=ON
BUILD_WITH_JULIA=OFF  # Requires Julia is installed
BUILD_WITH_ROSHAMBO=OFF  # Can't redistribute: unlicensed
BUILD_WITH_XINXIN=OFF  # Can't re-distribute: GPLv3
BUILD_WITH_LIBTORCH=OFF
BUILD_WITH_TENSORFLOW_CC=OFF
BUILD_WITH_GAMUT=OFF
BUILD_WITH_ORTOOLS=OFF
```

@michalsustr would you prefer to not include the public states at this point? I note that it is still marked as OFF in `global_variables.sh`; my understanding was that you had gotten public-tree CFR already working before you finished your internship. Shall we enable it?

@jhtschultz @finbarrtimbers tagging you to get your opinion on including `universal_poker`.

**Comment by finbarrtimbers** [comment]
Re: `universal_poker`, all the known bugs are fixed in it, and it has a decent amount of tests verifying that it's correct. So it should be fine to include it. It might be worth adding a disclaimer to the documentation saying that it hasn't been reviewed extensively by the OpenSpiel team and that we can't verify that it's correct.

**Comment by michalsustr** [comment]
> @michalsustr would you prefer to not include the public states at this point?

I would prefer not to include it. I don't know if someone is using that code at the moment, but the way how my research is oriented these days means I likely won't be using that particular code in the future.

> BUILD_WITH_ORTOOLS=OFF

Ortools has the same license as OpenSpiel (Apache-2.0), why not include it? Maybe now there are not too many algorithms that use it, but I imagine that there will be a number of LPs that will in the future.

**Comment by lanctot** [comment]
Ok, thanks about the public states API.

About OR-Tools: we currently don't support building it from source (we removed it, remember? In fact, wasn't it you who removed it? IIRC you tried it with v8 and it no longer worked). But even if we did: it's quite large, and would really add too much time to the build for a small fraction of users. Already the build time will be 1-2 minutes, which is long for a pip package. For people who want to use OR-Tools, they can build from source for now; we can probably build in support for dyanmically-linked loading of OR-Tools once we do a binary distribution.

**Comment by lanctot** [comment]
@michalsustr that means should I set `BUILD_WITH_EIGEN=OFF` too? Or was there other stuff that could use Eigen?

**Comment by michalsustr** [comment]
> we currently don't support building it from source

Ah ok I missed that, I didn't know you have to build everything from source for pip.

> BUILD_WITH_EIGEN=OFF too? Or was there other stuff that could use Eigen?

Yes it can be off, I don't think there was code that used it.

**Comment by lanctot** [comment]
> Ah ok I missed that, I didn't know you have to build everything from source for pip.

You don't have to but in our case it will be a fair bit of work to do the binary distribution properly (see reasons above) so we are starting with the sdist option.

**Comment by findmyway** [comment]
> will we still be able to register the Julia package as you were planning to? Do the release plans above affect this in any way, i.e. should we take this into account with the release planning? Also, I don't think you ever did submit the PR to modify the version as stated in this comment: #409 (comment) . Is the plan still the same?

Hi @lanctot , there were some changes to the underlying JLL (Julia Library) platform and I was supposed to make the related modifications after the upcoming next LTS release of Julia (@v1.6), which is estimated to be released in Dec. The release of OpenSpiel will not affect this process so you can go ahead with the plan first.

**Comment by lanctot** [comment]
I have successfully installed OpenSpiel from the PyPI test site (on Ubuntu 20.04):

```
lanctot@ladel:/tmp/open_spiel$ python3 -m pip install --index-url https://test.pypi.org/simple/ --no-deps open_spiel==0.2.0rc3
Looking in indexes: https://test.pypi.org/simple/
Collecting open_spiel==0.2.0rc3
  Downloading https://test-files.pythonhosted.org/packages/7b/66/e2ad0f6c83bebab780b70c9cb85268c50226dc38a36fd501f59d1a2f2ee8/open_spiel-0.2.0rc3.tar.gz (2.5 MB)
     |████████████████████████████████| 2.5 MB 1.4 MB/s
Building wheels for collected packages: open-spiel
  Building wheel for open-spiel (setup.py) ... done
  Created wheel for open-spiel: filename=open_spiel-0.2.0rc3-cp38-cp38-linux_x86_64.whl size=5655131 sha256=8b3e82f767f2652f7a9b7cbf769a4e79b0dbf028d0fe344581e169980c1e41e0
  Stored in directory: /home/lanctot/.cache/pip/wheels/56/7c/15/44b263d3f0adce5c9c483cc5e359a77c57b6d3bf7bd81af46f
Successfully built open-spiel
Installing collected packages: open-spiel
Successfully installed open-spiel-0.2.0rc3
lanctot@ladel:/tmp/open_spiel$ python3
Python 3.8.5 (default, Jul 28 2020, 12:59:40)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyspiel
>>> game = pyspiel.load_game("tic_tac_toe")
>>> state = game.new_initial_state()
>>> print(state)
...
...
...
>>>
```

**Comment by lanctot** [comment]
Ok, progress on a pip sdist is going quite well. I have to call out @calebho, he did 99% of the work on getting this setup. 

The last bits we're now waiting on for release 0.2.0 is:

- Bits and pieces needed for the pip sdist: https://github.com/deepmind/open_spiel/pull/448
- Tensorflow release 2.4.0 (any day now) so there's now security issues: https://github.com/deepmind/open_spiel/pull/445
- And I'd like to get Sheriff in, if possible: https://github.com/deepmind/open_spiel/pull/445

I expect we can issue the release next week or the week after.

**Comment by lanctot** [comment]
OpenSpiel 0.2.0 has been released, and it's now installable via pip! 

Release: https://github.com/deepmind/open_spiel/releases/tag/v0.2.0
Installation instructions for pip (usual place): https://github.com/deepmind/open_spiel/blob/master/docs/install.md

Enjoy!

**Comment by findmyway** [comment]
The latest progress of the Julia binary package of `v0.2.0`: https://github.com/JuliaPackaging/Yggdrasil/pull/2325
Once it gets resolved and the `Julia@v1.6 ` released, I'll make a PR to update the [install.sh](https://github.com/deepmind/open_spiel/blob/b8c2ff8e9a4f5dad9b179217f740ddb0df967f7c/open_spiel/scripts/install.sh#L192) and the [doc](https://github.com/deepmind/open_spiel/blob/master/docs/julia.md).

**Comment by lanctot** [comment]
Just wanted to wrap up this thread by saying that I still intend to look into bdist wheels for future versions, hopefully this year (maybe summer).
